File: 225-implement-stack-using-queues/implement-stack-using-queues.js

var MyStack = function() {
    this.q1=[];
    this.q2=[];
    
};

/** 
 * @param {number} x
 * @return {void}
 */
MyStack.prototype.push = function(x) {
    this.q1.push(x);
    
};

/**
 * @return {number}
 */
MyStack.prototype.pop = function() {
    let n=this.q1.length;
    for(let i=0;i<n-1;i++){
        this.q2.push(this.q1.shift());
    }
    let element = this.q1.shift();
    let temp = this.q1;
    this.q1 = this.q2;
    this.q2=temp;

    return element;

    
};

/**
 * @return {number}
 */
MyStack.prototype.top = function() {
      let n=this.q1.length;
    for(let i=0;i<n-1;i++){
        this.q2.push(this.q1.shift());
    }
    let element = this.q1.shift();
    this.q2.push(element);
    let temp = this.q1;
    this.q1 = this.q2;
    this.q2=temp;

    return element;
    
};

/**
 * @return {boolean}
 */
MyStack.prototype.empty = function() {
    return this.q1.length==0;
};

/** 
 * Your MyStack object will be instantiated and called as such:
 * var obj = new MyStack()
 * obj.push(x)
 * var param_2 = obj.pop()
 * var param_3 = obj.top()
 * var param_4 = obj.empty()
 */

File: 225-implement-stack-using-queues/implement-stack-using-queues.java
import java.util.LinkedList;
import java.util.Queue;

class MyStack {
    Queue<Integer> q1;
    Queue<Integer> q2;

    public MyStack() {
        this.q1 = new LinkedList<>();
        this.q2 = new LinkedList<>();
    }
    
    public void push(int x) {
        q1.add(x);
    }
    
    public int pop() {
        int size = q1.size();
        
        for (int i = 0; i < size - 1; i++) {
            q2.add(q1.poll());
        }
        
        int item = q1.poll();   // last element (stack top)
        
        // swap queues
        Queue<Integer> temp = q1;
        q1 = q2;
        q2 = temp;
        
        return item;
    }
    
    public int top() {
        int size = q1.size();
        
        for (int i = 0; i < size - 1; i++) {
            q2.add(q1.poll());
        }
        
        int item = q1.poll();
        q2.add(item);          // keep the top element in the queue
        
        // swap queues
        Queue<Integer> temp = q1;
        q1 = q2;
        q2 = temp;
        
        return item;
    }
    
    public boolean empty() {
        return q1.isEmpty();
    }
}
